src/libraries/System.Private.CoreLib/src/System/Resources/RuntimeResourceSet.cs

```diff
@@ -309,158 +269,104 @@ private IDictionaryEnumerator GetEnumeratorHelper()
         {
             if (key == null)
                 throw new ArgumentNullException(nameof(key));
-            if (Reader == null || _resCache == null)
+
+            ResourceReader? reader = _defaultReader;
+            Dictionary<string, ResourceLocator>? cache = _resCache;
+            Dictionary<string, ResourceLocator>? caseInsensitiveTable = _caseInsensitiveTable;
+            if (reader is null || cache is null)
                 throw new ObjectDisposedException(null, SR.ObjectDisposed_ResourceSet);
 
-            object? value = null;
-            ResourceLocator resLocation;
+            object? value;
+            ResourceLocator resEntry;
 
-            lock (Reader)
+            lock (cache)
```

Comment:
I think we need to get some perf numbers after changing this locking. The scope of locking with the cache is now changed and I don't know how much this can effect the behavior.

Reply:
What do you mean? The lock before locked about 120 lines whereas now it locks about 20 lines with early exits.

Reply:
I means changing the lock from the reader to the cache is changing the old locking scope even if we are locking later. I guess you can have multiple different readers pointing at the same cache. The old lock using the cache later was executing very small block and the code doesn't have always reach this lock. now you are locking with cache which can affect any readers using this cache and all readers will be blocked. I am just trying to ensure we are not regressing anything here including the performance.

Reply:
I'm could be missing something here but it does not matter which instance is used, all readers will be blocked if one enters the lock. If you are referring to this lock https://github.com/dotnet/runtime/blob/e7204f5d6fcaca5e097ec854b3be6055229fc442/src/libraries/System.Private.CoreLib/src/System/Resources/RuntimeResourceSet.cs#L358 then that's a nested lock which would be entered without outer lock

Reply:
> I'm could be missing something here but it does not matter which instance is used, all readers will be blocked if one enters the lock. If you are referring to this lock

That is my point, before your changes not all readers will be blocked. only the used reader instance will be blocked. now after your change, all readers sharing such cache will be blocked I guess.

Reply:
> before your changes not all readers will be blocked.

I'm still confused by what you mean by that. There was huge lock across the whole method which would block all concurrent readers (https://github.com/dotnet/runtime/blob/e7204f5d6fcaca5e097ec854b3be6055229fc442/src/libraries/System.Private.CoreLib/src/System/Resources/RuntimeResourceSet.cs#L318). The fact there was nested lock made it only worse not better.

Reply:
What I meant is, we have the following constructor. 

```C#
internal RuntimeResourceSet(IResourceReader reader)
```

Which take the reader as input. That means you can create multiple resource sets with the same reader if needed. Before your change, the lock was taken on the reader which means it can synchronize between all instances created with the same reader. After your change, the synchronization would happen using the cache which is created per one resource set object. That is the difference I am trying to point at. I don't know exactly how this can affect the behavior or the perf. 

The other side point, do you know if we ever compile with `#if RESOURCES_EXTENSIONS`? I am asking because this constructor is used only under this define. And I couldn't find who is using such constructor at all. I assume it was there for a good reason but I have no idea where or when it get used.

Reply:
I actually noticed that code is not needed anymore. Cleaned it up which should resolve your questions as well